<launch>

  <arg name="sensorOffsetX" default="0"/>
  <arg name="sensorOffsetY" default="0"/>
  <arg name="cameraOffsetZ" default="0"/>
  <arg name="twoWayDrive" default="true"/>
  <arg name="maxSpeed" default="1.0"/>
  <arg name="autonomyMode" default="true"/>
  <arg name="autonomySpeed" default="2.0"/>
  <arg name="joyToSpeedDelay" default="2.0"/>
  <arg name="goalX" default="0"/>
  <arg name="goalY" default="0"/>
  <arg name="reach_goal_thre_g" default="0.5"/>
  <arg name="points_topic" default="/fast_limo/deskewed" />
  <arg name="transform_points" default="true" />
  <arg name="points_target_frame" default="body" />
  <arg name="state_topic" default="/localization" />
  <arg name="odom_topic" default="/localization" />

  <!-- local planner (path generation) -->
  <node pkg="local_planner" type="localPlanner" name="localPlanner" output="screen" required="true">
    <param name="points_topic" value="$(arg points_topic)" />
    <param name="transform_points" value="$(arg transform_points)" />
    <param name="points_target_frame" value="$(arg points_target_frame)" />
    <param name="state_topic" value="$(arg state_topic)" />

    <param name="pathFolder" type="string" value="$(find local_planner)/paths" />
    <param name="vehicleLength" type="double" value="0.5" />
    <param name="vehicleWidth" type="double" value="1.0" />
    <param name="sensorOffsetX" value="$(arg sensorOffsetX)" />
    <param name="sensorOffsetY" value="$(arg sensorOffsetY)" />
    <param name="twoWayDrive" value="$(arg twoWayDrive)" />
    <param name="laserVoxelSize" type="double" value="0.05" />
    <param name="terrainVoxelSize" type="double" value="0.2" />
    <param name="useTerrainAnalysis" type="bool" value="false" />
    <param name="checkObstacle" type="bool" value="true" />
    <param name="checkRotObstacle" type="bool" value="false" />
    <param name="adjacentRange" type="double" value="5.0" />
    <param name="obstacleHeightThre" type="double" value="0.35" />
    <param name="groundHeightThre" type="double" value="0.05" />
    <param name="costHeightThre" type="double" value="0.1" />
    <param name="costScore" type="double" value="0.02" />
    <param name="useCost" type="bool" value="false" />
    <param name="usePathHysteresis" type="bool" value="true" />
    <param name="pathHysteresisRatio" type="double" value="0.15" />
    <param name="pathHysteresisDelta" type="double" value="0.05" />
    <param name="usePathSmoothing" type="bool" value="true" />
    <param name="pathSmoothAlpha" type="double" value="0.4" />
    <param name="pointPerPathThre" type="int" value="2" />
    <param name="minRelZ" type="double" value="-0.5" />
    <param name="maxRelZ" type="double" value="0.30" />
    <param name="maxSpeed" value="$(arg maxSpeed)" />
    <param name="dirWeight" type="double" value="0.02" />
    <param name="dirThre" type="double" value="90.0" />
    <param name="dirToVehicle" type="bool" value="false" />
    <param name="pathScale" type="double" value="1.25" />
    <param name="minPathScale" type="double" value="0.75" />
    <param name="pathScaleStep" type="double" value="0.25" />
    <param name="pathScaleBySpeed" type="bool" value="true" />
    <param name="minPathRange" type="double" value="1.0" />
    <param name="pathRangeStep" type="double" value="0.5" />
    <param name="pathRangeBySpeed" type="bool" value="true" />
    <param name="pathCropByGoal" type="bool" value="true" />
    <param name="autonomyMode" value="$(arg autonomyMode)" />
    <param name="autonomySpeed" value="$(arg autonomySpeed)" />
    <param name="joyToSpeedDelay" value="$(arg joyToSpeedDelay)" />
    <param name="joyToCheckObstacleDelay" type="double" value="5.0" />
    <param name="goalClearRange" type="double" value="0.5" />
    <param name="goalX" type="double" value="$(arg goalX)" />
    <param name="goalY" type="double" value="$(arg goalY)" />
    <!-- 末端对齐不在 localPlanner 中处理，仅用 reach_goal_thre_g 作为位置到达阈值 -->
    <param name="reach_goal_thre_g" type="double" value="$(arg reach_goal_thre_g)" />  <!-- 位置到达阈值 [m] -->
  </node>

  <!-- mpcc follower (tracking) -->
  <node pkg="local_planner" type="mpccfollower" name="mpccfollower" output="screen" required="true">
    <rosparam command="load" file="$(find local_planner)/config/mpccfollower.yaml" />
    <param name="odom_topic" value="$(arg odom_topic)" />
    <param name="maxSpeed" value="$(arg maxSpeed)" />
    <param name="autonomyMode" value="$(arg autonomyMode)" />
    <param name="autonomySpeed" value="$(arg autonomySpeed)" />
    <param name="joyToSpeedDelay" value="$(arg joyToSpeedDelay)" />
  </node>

  <!-- mpcc prediction visualization -->
  <node pkg="local_planner" type="mpcc_pred_visualizer.py" name="mpcc_pred_visualizer" output="screen">
    <param name="pred_topic" value="/mpcc_pred_path"/>
    <param name="odom_topic" value="/localization"/>
    <param name="marker_topic" value="/mpcc_pred_marker"/>
  </node>

</launch>
